File: app/src/main/java/com/qurantv/app/data/repo/QuranTextRepository.kt
package com.qurantv.app.data.repo

import android.content.Context
import com.qurantv.app.data.api.QuranComApi
import com.qurantv.app.data.api.VersesResponse
import com.qurantv.app.data.cache.JsonDiskCache
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json

/**
 * Authentic Uthmani Quran text (decision D2):
 *  - primary: bundled Tanzil `quran-uthmani.txt` asset (canonical, offline),
 *    parsed once into an in-memory `verse_key -> text` map (~6k entries);
 *  - fallback: Quran.com API v4 `text_uthmani`, cached per surah on disk.
 */
class QuranTextRepository(
    private val context: Context,
    private val quranApi: QuranComApi,
    private val cache: JsonDiskCache,
    private val json: Json,
) {

    @Volatile
    private var tanzilMap: Map<String, String>? = null

    private suspend fun loadTanzil(): Map<String, String> {
        tanzilMap?.let { return it }
        return withContext(Dispatchers.IO) {
            val map = try {
                context.assets.open("quran/quran-uthmani.txt").bufferedReader(Charsets.UTF_8).useLines { lines ->
                    buildMap {
                        lines.forEach { line ->
                            val parts = line.split('|', limit = 3)
                            if (parts.size == 3) {
                                val key = "${parts[0]}:${parts[1]}"
                                put(key, parts[2])
                            }
                        }
                    }
                }
            } catch (e: Exception) {
                emptyMap()
            }
            tanzilMap = map
            map
        }
    }

    suspend fun verseText(surahId: Int, verseNumber: Int): String? {
        val key = "$surahId:$verseNumber"
        val raw = loadTanzil()[key]
        if (raw != null) {
            // Tanzil prefixes the basmala to verse 1 of surahs 2–114; the recitation
            // recites it as the surah header, not as part of verse 1 — strip it so the
            // displayed text matches the audio exactly. The prefix is taken from the
            // data itself (1|1) to guarantee an exact character match.
            if (surahId in 2..114 && verseNumber == 1) {
                val basmala = loadTanzil()["1:1"] ?: return raw
                return stripBasmala(raw, basmala)
            }
            return raw
        }
        return cachedVerseFromQuranCom(surahId, verseNumber, key)
    }

    private fun stripBasmala(text: String, basmala: String): String {
        if (!text.startsWith(basmala)) return text.trim()
        val rest = text.removePrefix(basmala).trimStart()
        return rest.ifEmpty { text.trim() }
    }

    private suspend fun cachedVerseFromQuranCom(surahId: Int, verseNumber: Int, key: String): String? {
        val cacheKey = "ch$surahId"
        return cache.singleFlight(cacheKey) {
            val cached = cache.read(JsonDiskCache.QURAN_TEXT, cacheKey)
            val verses = if (cached != null) {
                json.decodeFromString<VersesResponse>(cached).verses
            } else {
                try {
                    quranApi.versesUthmani(surahId).also { list ->
                        runCatching {
                            cache.write(JsonDiskCache.QURAN_TEXT, cacheKey, json.encodeToString(VersesResponse(list)))
                        }
                    }
                } catch (e: Exception) {
                    emptyList()
                }
            }
            verses.firstOrNull { it.verseKey == key }?.textUthmani
        }
    }
}

File: web/src/data/repo/QuranTextRepository.ts
// Ported 1:1 from app/src/main/java/com/qurantv/app/data/repo/QuranTextRepository.kt
// Authentic Uthmani Quran text:
//  - primary: bundled Tanzil `quran-uthmani.txt` (canonical, offline), parsed
//    once into an in-memory `verse_key -> text` map (~6k entries);
//  - fallback: Quran.com API v4 `text_uthmani`, cached per surah forever.

import { QuranComApi } from "../api/QuranComApi";
import type { VerseDto } from "../api/Dtos";
import { CACHE, JsonDiskCache } from "../cache/JsonDiskCache";
import { loadAssetText } from "../assetLoader";

const TANZIL_ASSET = "quran/quran-uthmani.txt";

export class QuranTextRepository {
  private tanzilMap: Map<string, string> | null = null;

  constructor(
    private readonly quranApi: QuranComApi,
    private readonly cache: JsonDiskCache,
  ) {}

  private async loadTanzil(): Promise<Map<string, string>> {
    if (this.tanzilMap !== null) return this.tanzilMap;
    const map = new Map<string, string>();
    try {
      // XHR not fetch: file:// in the Android TV webview blocks fetch().
      const text = await loadAssetText(TANZIL_ASSET);
      for (const line of text.split("\n")) {
        const parts = line.split("|", 3);
        if (parts.length === 3 && parts[0] !== "" && parts[1] !== "") {
          map.set(`${parts[0]}:${parts[1]}`, parts[2]);
        }
      }
    } catch {
      // fall back to Quran.com per-verse lookups
    }
    this.tanzilMap = map;
    return map;
  }

  /** Text length of a verse (0 when unknown) — used by the page highlight estimator. */
  async verseTextLength(surahId: number, verseNumber: number): Promise<number> {
    return (await this.loadTanzil()).get(`${surahId}:${verseNumber}`)?.length ?? 0;
  }

  async verseText(surahId: number, verseNumber: number): Promise<string | null> {
    const key = `${surahId}:${verseNumber}`;
    const raw = (await this.loadTanzil()).get(key);
    if (raw !== undefined) {
      // Tanzil embeds the basmala as a prefix of verse 1 for surahs 2–114; the
      // recitation recites it as its own segment (timing index 0, shown as the
      // basmala header), NOT as part of verse 1 — strip the prefix so displayed
      // text matches the audio 1:1. The prefix comes from the data itself (1|1).
      // Surah 9 has no basmala, and surah 1's verse 1 IS the basmala — both are
      // naturally untouched by this branch.
      if (surahId >= 2 && surahId <= 114 && verseNumber === 1) {
        const basmala = (await this.loadTanzil()).get("1:1");
        if (basmala !== undefined) {
          return stripBasmala(raw, basmala);
        }
      }
      return raw;
    }
    return this.cachedVerseFromQuranCom(surahId, key);
  }

  private async cachedVerseFromQuranCom(
    surahId: number,
    key: string,
  ): Promise<string | null> {
    const cacheKey = `ch${surahId}`;
    return this.cache.singleFlight(cacheKey, async () => {
      const cached = await this.cache.read(CACHE.QURAN_TEXT, cacheKey);
      let verses: VerseDto[];
      if (cached !== null) {
        verses = (JSON.parse(cached) as { verses: VerseDto[] }).verses;
      } else {
        try {
          verses = await this.quranApi.versesUthmani(surahId);
          await this.cache.write(CACHE.QURAN_TEXT, cacheKey, JSON.stringify({ verses }));
        } catch {
          verses = [];
        }
      }
      return verses.find((v) => v.verse_key === key)?.text_uthmani ?? null;
    });
  }
}

export function stripBasmala(text: string, basmala: string): string {
  if (!text.startsWith(basmala)) return text.trim();
  const rest = text.slice(basmala.length).trimStart();
  return rest.length > 0 ? rest : text.trim();
}
